### ISSUE TITLE
"Could not send eof data to FOC at the other end: std::bad_alloc" during volume failover with 10ms latency.

### ISSUE THREAD

**Issue by kinvaris** [openvstorage/volumedriver#19]
We've got a setup with 2 datacenters, between those 2 datacenters there is 10ms latency. The DTL is configured across the datacenters. This would mean if 1 DC fails, the other one should take over.

While performing the test (shutting down one compute node in DC2) we experienced these issue's with the volumedriver during failover (in DC1):
```
2016-05-19 15:13:51:661917 +0200 CEST -- error -- FuseInterface -- convert_exceptions: "/POOL_3577/DISK_2119.raw": caught fungi::IOException: Problem with failover cache Socket::read/closed , code 49 -- [0x00007fc772ffb700]
2016-05-19 15:13:51:921902 +0200 CEST -- error -- FuseInterface -- convert_exceptions: "/POOL_3577/DISK_2119.raw": Volume location not found 74027914-50b4-426b-bfc2-1902758c7475 -- [0x00007fc8067fc700]
2016-05-19 15:14:08:416886 +0200 CEST -- error -- FuseInterface -- convert_exceptions: "/POOL_3577/DISK_5134.raw": caught fungi::IOException: Problem with failover cache Socket::read/closed , code 738197656 -- [0x00007fc772ffb700]
2016-05-19 15:14:08:677810 +0200 CEST -- error -- FuseInterface -- convert_exceptions: "/POOL_3577/DISK_5134.raw": Volume location not found 40b18dd4-fed3-4d62-baa4-41d077ce7074 -- [0x00007fc8257fa700]
2016-05-19 15:19:23:120761 +0200 CEST -- error -- FuseInterface -- convert_exceptions: "/POOL_3577/DISK_2119.raw": Volume location not found 74027914-50b4-426b-bfc2-1902758c7475 -- [0x00007fc805ffb700]
2016-05-19 15:19:23:379260 +0200 CEST -- error -- FuseInterface -- convert_exceptions: "/POOL_3577/DISK_5134.raw": Volume location not found 40b18dd4-fed3-4d62-baa4-41d077ce7074 -- [0x00007fc8257fa700]
```

After some digging, we see that one volume was successfully stolen but the other 2 failed due to a problem with the DTL: (steal successfully executed here)
```
2016-05-19 15:02:46:491421 +0200 CEST -- notice -- VolManager -- Backend Restart, Namespace: 96a5760d-8188-4f07-9f96-dd628474c73f, owner tag 100, FINISHED -- [
0x00007fc805ffb700]
2016-05-19 15:02:46:491131 +0200 CEST -- info -- BackendInterface -- handle_eventual_consistency_: 47b19a03-e315-4e03-8981-f340f3377d9b: caller wants latest ve
rsion: InsistOnLatestVersion::F, backend offers strict consistency: false -> not insisting on latest version towards the backend -- [0x00007fc7c4ffe700]
2016-05-19 15:02:46:491511 +0200 CEST -- info -- BackendConnectionInterfaceLogger -- Logger: Entering partial_read 47b19a03-e315-4e03-8981-f340f3377d9b -- [0x0
0007fc7c4ffe700]
2016-05-19 15:02:46:491142 +0200 CEST -- info -- BackendInterface -- handle_eventual_consistency_: 47b19a03-e315-4e03-8981-f340f3377d9b: caller wants latest ve
rsion: InsistOnLatestVersion::F, backend offers strict consistency: false -> not insisting on latest version towards the backend -- [0x00007fc7737fc700]
2016-05-19 15:02:46:491588 +0200 CEST -- info -- BackendConnectionInterfaceLogger -- Logger: Entering partial_read 47b19a03-e315-4e03-8981-f340f3377d9b -- [0x0
0007fc7737fc700]
2016-05-19 15:02:46:503245 +0200 CEST -- info -- BackendConnectionInterfaceLogger -- ~Logger: Exiting partial_read for 47b19a03-e315-4e03-8981-f340f3377d9b -- 
[0x00007fc7737fc700]
2016-05-19 15:02:46:503330 +0200 CEST -- info -- VFSObjectRouter -- steal_: 96a5760d-8188-4f07-9f96-dd628474c73f: successfully stolen from myvpool-ar8v6v2CxjWe
UuO1V -- [0x00007fc805ffb700]
2016-05-19 15:02:46:503362 +0200 CEST -- info -- VFSObjectRouter -- maybe_steal_: 96a5760d-8188-4f07-9f96-dd628474c73f: retrying I/O after stealing -- [0x00007
fc805ffb700]
```

As you can see: (failed steal)
```
2016-05-19 15:13:51:661753 +0200 CEST -- notice -- VolManager -- Backend Restart, Namespace: 74027914-50b4-426b-bfc2-1902758c7475, owner tag 111, EXCEPTION -- 
[0x00007fc772ffb700]
2016-05-19 15:13:51:661799 +0200 CEST -- error -- VFSLocalNode -- restart_volume_from_backend_: 74027914-50b4-426b-bfc2-1902758c7475: failed to restart volume:
 : Problem with failover cache Socket::read/closed  -- [0x00007fc772ffb700]
2016-05-19 15:13:51:661849 +0200 CEST -- error -- VFSObjectRouter -- steal_: 74027914-50b4-426b-bfc2-1902758c7475: failed to restart: Problem with failover cac
he Socket::read/closed  -- [0x00007fc772ffb700]
2016-05-19 15:13:51:661917 +0200 CEST -- error -- FuseInterface -- convert_exceptions: "/POOL_3577/DISK_2119.raw": caught fungi::IOException: Problem with failover cache Socket::read/closed , code 49 -- [0x00007fc772ffb700]
```

Looking at the DTL:
```
2016-05-19 15:13:51:188925 +0200 CEST -- info -- DtlFileBackend -- get_entries: 74027914-50b4-426b-bfc2-1902758c7475: processing SCO 00_00001efd_00 -- [0x00007fc7667fc700]
2016-05-19 15:13:51:195067 +0200 CEST -- warning -- FailOverCacheProtocol -- operator(): Could not send eof data to FOC at the other end: std::bad_alloc - ignored -- [0x00007fc7667fc700]
2016-05-19 15:13:51:195115 +0200 CEST -- error -- FailOverCacheProtocol -- run: Exception in thread: std::bad_alloc -- [0x00007fc7667fc700]

2016-05-19 15:14:08:301230 +0200 CEST -- info -- DtlFileBackend -- get_entries: 40b18dd4-fed3-4d62-baa4-41d077ce7074: processing SCO 00_00001d69_00 -- [0x00007
f5be23e9700]
2016-05-19 15:14:08:311142 +0200 CEST -- warning -- FailOverCacheProtocol -- operator(): Could not send eof data to FOC at the other end: std::bad_alloc - igno
red -- [0x00007f5be23e9700]
```

**Comment by redlicha** [comment]
PR: https://github.com/openvstorage/volumedriver/pull/23

**Comment by redlicha** [comment]
Fix was merged into support/5.6.x and dev branches.

**Comment by redlicha** [comment]
New dev packages including a fix were uploaded to unstable.

**Comment by wimpers** [comment]
@redlicha can this one be closed?

**Comment by redlicha** [comment]
Fixed in support/5.6.x (still needs to be packaged) and the dev code, including the 6.0.0 release.
